dali/kernels/signal/dct/dct_cpu.cc

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/signal/dct/dct_cpu.h"
+#include <cmath>
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/common/for_axis.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace dct {
+
+namespace {
+
+template <typename T>
+void FillCosineTableTypeI(T *table, int64_t input_length, int64_t ndct, bool normalize) {
+  assert(input_length > 1);
+  assert(!normalize);
+  T phase_mul = M_PI / (input_length - 1);
+  int64_t idx = 0;
+  for (int64_t k = 0; k < ndct; k++) {
+    table[idx++] = T(0.5);  // n = 0
+    for (int64_t n = 1; n < input_length-1; n++) {
+      table[idx++] = std::cos(phase_mul * k * n);
+    }
+    table[idx++] = k % 2 == 0 ?  T(0.5) : -T(0.5);  // n = input_length - 1
+  }
+}
+
+template <typename T>
+void FillCosineTableTypeII(T *table, int64_t input_length, int64_t ndct, bool normalize) {
+  T phase_mul = M_PI / input_length;
+  T factor_k_0 = 1, factor_k_i = 1;
+  if (normalize) {
+    factor_k_i = std::sqrt(2 / T(input_length));
+    factor_k_0 = factor_k_i / std::sqrt(T(2));
+  }
+  int64_t idx = 0;
+  for (int64_t k = 0; k < ndct; k++) {
+    T norm_factor = (k == 0) ? factor_k_0 : factor_k_i;
+    for (int64_t n = 0; n < input_length; n++) {
+      table[idx++] = norm_factor * std::cos(phase_mul * (n + T(0.5)) * k);
```

Comment:
If it's a one-off calculation, then the code below give more precision:
```suggestion
      table[idx++] = T(norm_factor * std::cos(phase_mul * (n + 0.5) * k);
```